### ISSUE TITLE
Method Implementation Requirements for Solvers

### ISSUE THREAD

**Issue by etotheipluspi** [JuliaPOMDP/POMDPs.jl#91]
Right now we have no standardized way of telling the user what methods they need to implement (other than putting a list of them in the docs). I added a utility tool to POMDPs.jl that you can call like this:

```julia
POMDPs.get_methods(function_list)
```
Where `function_list` is a vector of function names that looks something like `function_list=[states, pdf, ...]`. 

I think it may be useful to have each solver implement something like a `required()` function that prints the methods that need to be implemented and their input arguments. See how I did this in [QMDP](https://github.com/JuliaPOMDP/QMDP.jl/blob/master/src/required.jl) using the above function. The user can do something like:

```julia
using QMDP
QMDP.required_methods()
```

Output:

```
8-element Array{Method,1}:
 states{S,A}(problem::Union{POMDPs.MDP{S,A},POMDPs.POMDP{S,A,O}}, state::S, sts::POMDPs.AbstractSpace{S})                               
 actions{S,A}(problem::Union{POMDPs.MDP{S,A},POMDPs.POMDP{S,A,O}}, state::S, acts::POMDPs.AbstractSpace{A})                             
 iterator{T}(s::POMDPs.AbstractSpace{T})                                                                                                
 create_transition_distribution(problem::Union{POMDPs.MDP{S,A},POMDPs.POMDP{S,A,O}})                                                    
 transition{S,A}(problem::Union{POMDPs.MDP{S,A},POMDPs.POMDP{S,A,O}}, state::S, action::A, distribution::POMDPs.AbstractDistribution{S})
 reward{S,A}(problem::Union{POMDPs.MDP{S,A},POMDPs.POMDP{S,A,O}}, state::S, action::A)                                                  
 pdf{T}(d::POMDPs.AbstractDistribution{T}, x::T)                                                                                        
 state_index{S,A}(problem::Union{POMDPs.MDP{S,A},POMDPs.POMDP{S,A,O}}, s::S)      
```

The way the method array gets generated is a bit awkward, so if you have a way of making it better please check out the function in utils.jl.

TLDR; Please add a `Solver.required` method to your solver that calls `POMDPs.get_methods` on a vector of required functions for your solver. This will allow the user to check what functions they need to implement from the REPL. 

Edit: changed `required()` to `required_methods()` for future reference.

**Comment by zsunberg** [comment]
Good idea. Should we perhaps name it something more specific like "required_methods"?

**Comment by etotheipluspi** [comment]
`required_methods ` is a better name, since `required` can also refer to dependencies. 

Unless anyone has objections to that name, I'm going to change it, and tag the current version of POMDPs.jl + submit a PR to METADATA.

**Comment by ebalaban** [comment]
Sounds good to me. The function name is fine too.

**Comment by zsunberg** [comment]
Sounds good

**Comment by etotheipluspi** [comment]
Closing this. The convention is to have `SolverName.required_methods()` call that returns a vector of methods used by the solver.
